kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,389 @@
+# RFC: A(nother) Uniform Analyzer Interface for Kythe
+
+https://github.com/kythe/kythe/issues/3306.md
+
+This document describes an operational protocol to implement communication
+between a Kythe analyzer and a driver program using a pair of reliable simplex
+byte streams such as Unix pipes.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Background and Motivation
+
+The [Kythe](http://kythe.io) project supports a pluggable language-agnostic
+ecosystem of tools for analyzing source code. Pluggability is provided by a
+standard interface that a tool must implement in order to qualify as a Kythe
+analyzer.
+
+At a high level, an _analyzer_ is a program that consumes Kythe compilation
+records as input, and emits `kythe.proto.Entry` protocol buffer messages as
+output. However, the base definition of Kythe did not specify how this interface
+is to be implemented concretely. That leaves the implementor a great deal of
+flexibility, but makes it unnecessarily difficult to integrate analyzers that
+are implemented in different languages.
+
+The purpose of this proposal is to specify an explicit wire-level protocol for
+communicating compilation records to, and collecting entries from, a Kythe
+analyzer. Some important goals of this proposal are:
+
+1.  To decouple analyzer implementations from the storage representations of
+    compilation records and entry messages. This avoids the need to update
+    analyzers when storage formats change, and makes it more practical to run
+    analyzers in a sandbox.
+
+2.  To make it easier to run analyzers as standalone tools, and reduce or
+    eliminate the need to implement separate wrapper tools for testing vs.
+    production.
+
+3.  To reduce friction for analyzers written in languages that do not have good
+    support for efficient manipulation of binary data, or which require FFI to
+    get good performance.
+
```

Comment:
So something I just realized about this proposal: it essentially prohibits an analyzer from performing concurrent analyses. The core of the reason is that the file system virtualization is not sensitive to which analysis is executing. It's definitely possible to change the virtualized file data in between sequential analyses, but concurrent analyses would see each others' input files. In the case where a generated file is generated from two different build configs and has different content, we'd have a correctness issue. Even outside that case, if the virtualization synthesizes directory listings, that could also cause problems. I suspect there are other potential issues here.

Is it worth while giving up on concurrent analyses?

Reply:
> Is it worth while giving up on concurrent analyses?

IMO yes. The existing driver only offers one analysis at a time to the analyzer, partly because in the past when we tried to do more we discovered many of the compiler libraries are not safe for concurrency (e.g., using static data and caches without locks). When I wrote #2966 I had in mind to allow the _analyzer_ to decide, but I don't actually expect that feature to be used much in practice, and I think it could be sacrificed.